### ISSUE TITLE
Is there a conflict between class_weight and seg_ pad_val？

### ISSUE THREAD

**Issue by SquirrelEdison** [open-mmlab/mmsegmentation#3831]
I fine tuned the segnext model on a custom dataset (including 6 types of backgrounds) and found that the value of the label was normal in the loader (from 0 to num_classes-1), but when passed into the class_weight calculation, there was a sudden addition of 255 values in the label. After debugging, I found that these values were caused by seg_pad_val.

![image](https://github.com/user-attachments/assets/3921d981-9e12-48b6-a571-1e8a29c7a6b0)
label's val is error-free in the dataloader

![image](https://github.com/user-attachments/assets/769a4512-414f-4fb3-8085-362c5392c58b)
label's val is added some '255' val

and in class_weight calculation  (/conda/lib/python3.9/site-packages/mmsegmentation-1.2.2-py3.9.egg/mmseg/models/losses/cross_entropy_loss.py):
```python
if (avg_factor is None) and reduction == 'mean':
        if class_weight is None:
            for cls in label:
                print("cls: ", cls)
            if avg_non_ignore:
                avg_factor = label.numel() - (label
                                              == ignore_index).sum().item()
            else:
                avg_factor = label.numel()

        else:
            # the average factor should take the class weights into account
            # print("label: ", label)
            # print("label_shape: ", label.shape)
            # print("class_weight: ", class_weight)
            # print("cls in label:")
            # for cls in label:
            #         print("cls: ", cls)
            #         print("class_weight[cls]: ",class_weight[cls])
            label_weights = torch.stack([class_weight[cls] for cls in label
                                         ]).to(device=class_weight.device)

            if avg_non_ignore:
                label_weights[label == ignore_index] = 0
            avg_factor = label_weights.sum()
```
`label_weights = torch.stack([class_weight[cls] for cls in label
                                         ]).to(device=class_weight.device)`
the cls in class_weight[cls] will be '255', and this code will cause index out of bound 

when I changed the val of seg_pad_val from '255' to '0', the label's val was back to normal
![image](https://github.com/user-attachments/assets/d098179a-4e11-4394-aec9-e241ba551ff7)

`segnext model config file`:
```python
_base_ = [
    '../_base_/default_runtime.py', '../_base_/schedules/schedule_20k.py',
    '../_base_/datasets/fire6dataset.py'
]
# model settings
checkpoint_file = 'https://download.openmmlab.com/mmsegmentation/v0.5/pretrain/segnext/mscan_t_20230227-119e8c9f.pth'  # noqa
ham_norm_cfg = dict(type='GN', num_groups=32, requires_grad=True)
crop_size = (512, 512)
data_preprocessor = dict(
    type='SegDataPreProcessor',
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    bgr_to_rgb=True,
    pad_val=0,
    seg_pad_val=255,  
    size=(512, 512),
    test_cfg=dict(size_divisor=32))
model = dict(
    type='EncoderDecoder',
    data_preprocessor=data_preprocessor,
    pretrained=None,
    backbone=dict(
        type='MSCAN',
        init_cfg=dict(type='Pretrained', checkpoint=checkpoint_file),
        embed_dims=[32, 64, 160, 256],
        mlp_ratios=[8, 8, 4, 4],
        drop_rate=0.0,
        drop_path_rate=0.1,
        depths=[3, 3, 5, 2],
        attention_kernel_sizes=[5, [1, 7], [1, 11], [1, 21]],
        attention_kernel_paddings=[2, [0, 3], [0, 5], [0, 10]],
        act_cfg=dict(type='GELU'),
        norm_cfg=dict(type='BN', requires_grad=True)),
    decode_head=dict(
        type='LightHamHead',
        in_channels=[64, 160, 256],
        in_index=[1, 2, 3],
        channels=256,
        ham_channels=256,
        dropout_ratio=0.1,
        num_classes=6,  # include background
        norm_cfg=ham_norm_cfg,
        align_corners=False,
        loss_decode=[
        dict(type='CrossEntropyLoss', loss_name='loss_ce', use_sigmoid=False, loss_weight=1.0, class_weight=[0.87, 1.0, 1.85, 1.85, 1.17, 2.14]),
        # dict(type='CrossEntropyLoss', loss_name='loss_ce', use_sigmoid=False, loss_weight=1.0),
        dict(type='DiceLoss', loss_name='loss_dice', loss_weight=1.0)
            ],
        ham_kwargs=dict(
            MD_S=1,
            MD_R=16,
            train_steps=6,
            eval_steps=7,
            inv_t=100,
            rand_init=True)),
    # model training and testing settings
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))

# dataset settings
train_dataloader = dict(batch_size=8)

# optimizer
optim_wrapper = dict(
    _delete_=True,
    type='OptimWrapper',
    optimizer=dict(
        type='AdamW', lr=0.00006, betas=(0.9, 0.999), weight_decay=0.01),
    paramwise_cfg=dict(
        custom_keys={
            'pos_block': dict(decay_mult=0.),
            'norm': dict(decay_mult=0.),
            'head': dict(lr_mult=10.)
        }))

param_scheduler = [
    dict(
        type='LinearLR', start_factor=1e-6, by_epoch=False, begin=0, end=1500),
    dict(
        type='PolyLR',
        power=1.0,
        begin=1500,
        end=20000,
        eta_min=0.0,
        by_epoch=False,
    )
]
```
`_base_/datasets/fire6dataset.py`:
```python
# dataset settings
# reduce_zero_label = False
dataset_type = 'Fire6Dataset'
data_root = 'data/ade/FireData'
crop_size = (512, 512)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', reduce_zero_label=False),
    dict(
        type='RandomResize',
        scale=(2048, 512),
        ratio_range=(0.5, 2.0),
        keep_ratio=True),
    dict(type='RandomCrop', crop_size=crop_size, cat_max_ratio=0.75),
    dict(type='RandomFlip', prob=0.5),
    # dict(type='PhotoMetricDistortion'),
    dict(type='PackSegInputs')
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', scale=(2048, 512), keep_ratio=True),
    # add loading annotation after ``Resize`` because ground truth
    # does not need to do resize data transform
    dict(type='LoadAnnotations', reduce_zero_label=False),
    dict(type='PackSegInputs')
]
img_ratios = [0.5, 0.75, 1.0, 1.25, 1.5, 1.75]
tta_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(
        type='TestTimeAug',
        transforms=[
            [
                dict(type='Resize', scale_factor=r, keep_ratio=True)
                for r in img_ratios
            ],
            [
                dict(type='RandomFlip', prob=0., direction='horizontal'),
                dict(type='RandomFlip', prob=1., direction='horizontal')
            ], [dict(type='LoadAnnotations')], [dict(type='PackSegInputs')]
        ])
]
train_dataloader = dict(
    batch_size=4,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='InfiniteSampler', shuffle=True),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_prefix=dict(
            img_path='images/training', seg_map_path='annotations/training'),
        pipeline=train_pipeline))
val_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_prefix=dict(
            img_path='images/validation',
            seg_map_path='annotations/validation'),
        pipeline=test_pipeline))
test_dataloader = val_dataloader

val_evaluator = dict(type='IoUMetric', iou_metrics=['mIoU'])
test_evaluator = val_evaluator
```

`mmseg/datasets/fire_dataset.py`:
```python
# Copyright (c) OpenMMLab. All rights reserved.
from mmseg.registry import DATASETS
from .basesegdataset import BaseSegDataset


@DATASETS.register_module()
class FireDataset(BaseSegDataset):
    """ADE20K dataset.

    In segmentation map annotation for ADE20K, 0 stands for background, which
    is not included in 150 categories. ``reduce_zero_label`` is fixed to True.
    The ``img_suffix`` is fixed to '.jpg' and ``seg_map_suffix`` is fixed to
    '.png'.
    """
    METAINFO = dict(
        classes=('fire','smoke_black','smoke_white','smoke_yellow','spark'),
        palette=[[120, 120, 120], [180, 120, 120], [6, 230, 230], [80, 50, 50],
                 [4, 200, 3]])

    def __init__(self,
                 img_suffix='.jpg',
                 seg_map_suffix='.png',
                 reduce_zero_label=True,
                 **kwargs) -> None:
        super().__init__(
            img_suffix=img_suffix,
            seg_map_suffix=seg_map_suffix,
            reduce_zero_label=reduce_zero_label,
            **kwargs)
```

**Comment by yhxu022** [comment]
您好！我也遇到了相同的错误，但是我不希望将seg_pad_val设置为0，因为这会导致类别“0”的样本数大大增加。请问有没有办法可以使得seg_pad_val=255的同时，可以正常使用cls weight？

**Comment by SquirrelEdison** [comment]
> 您好！我也遇到了相同的错误，但是我不希望将seg_pad_val设置为0，因为这会导致类别“0”的样本数大大增加。请问有没有办法可以使得seg_pad_val=255的同时，可以正常使用cls weight？

抱歉，我也不知道了:(

**Comment by MyHogsmeade** [comment]
Meet the same problem, I use "seg_pad_val=0".

**Comment by AndreaSanchini** [comment]
Has anybody managed to fix this issue? I would want to avoid using "seg_pad_val=0"
